### PR TITLE
Add helper to run a containerized build environment

### PR DESCRIPTION
**Pull request by mweinelt**
Using `make container` or, if you don't have automake/gmake on your host
system, `./scripts/container.sh` will build an image for the current
branch your are on and drop you into a shell running inside a container
using that image.

From there all tooling required to work on Gluon is available.

Supports both docker and podman, though I have not tested docker. Looks like this:

```
[hexa@nyx:~/git/gluon/master]$ make container
STEP 1: FROM debian:buster-slim
STEP 2: ARG DEBIAN_FRONTEND=noninteractive
--> Using cache af97c833b3010a3ce2cb13273436785a26be46872e91106bd2b05ba3aedcadf8
--> af97c833b30
STEP 3: RUN apt-get update && apt-get install -y --no-install-recommends     ca-certificates     file     git     subversion     python3     build-essential     gawk     unzip     libncurses5-dev     zlib1g-dev     libssl-dev     libelf-dev     wget     rsync     time     qemu-utils     ecdsautils     lua-check     shellcheck   && rm -rf /var/lib/apt/lists/*
--> Using cache 15535e5b0a9a2962b55fb4486e5eb74c3b6d324acb6675663582bff2280e3634
--> 15535e5b0a9
STEP 4: RUN useradd -d /gluon gluon
--> Using cache e797f42852743e179ef6ed0a65fc33d6a3ba98c70a918c8c072e28c829adf52a
--> e797f428527
STEP 5: USER gluon
--> Using cache 0ee1195cf45d2eeb2f869edec7560a27ea94d2e54b7514a2d4f7be236fb1f932
--> 0ee1195cf45
STEP 6: VOLUME /gluon
--> Using cache 13241cc527d13e5bc48014fdffb7adee64dd759a9bc1147734f0d265be684d21
--> 13241cc527d
STEP 7: WORKDIR /gluon
--> Using cache 50c35838b238fb961522ba7ae30f6f304f08e317630a9a723892c451e6cbe7ed
STEP 8: COMMIT gluon:container-helper
--> 50c35838b23
Successfully tagged localhost/gluon:container-helper
Successfully tagged localhost/gluon:master
50c35838b238fb961522ba7ae30f6f304f08e317630a9a723892c451e6cbe7ed
gluon@4c0a2a065308:~$ 
```

<details><summary>Weird approach to run a script inside docker, that breaks stdin and won't let you exit :rofl:</summary>
Another funny use case could be

<pre>
cat <<EOF | ./scripts/container.sh
  make update
  make -j2 V=s GLUON_TARGET=ipq40xx-generic GLUON_RELEASE=testing
  exit
EOF
</pre>
</details>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -178,6 +178,10 @@ config: $(LUA) FORCE
 	$(GLUON_ENV) $(LUA) scripts/target_config_check.lua
 
 
+container: FORCE
+	@scripts/container.sh
```
Comment (NeoRaider) [inline Makefile:182]: Does it really make sense to put this in the Makefile, even though it is not part of the actual build?

I'd just put the script under `contrib/` to be called without the Makefile.

Comment (mweinelt) [reply Makefile:182]: I have no particular love for `make container` besides establishing good discoverability.

Setting up a build environment is likely as much part of the actual build as the linter steps.

**Comment by herbetom** [comment]
I think this PR is kinda missing a [3. topic: docs](https://github.com/freifunk-gluon/gluon/labels/3.%20topic%3A%20docs) tag.

Based on a quick test, it seems to work and is quite handy, even if initializing a container could be a bit faster ("Sending build context to Docker daemon"). :+1:

File: scripts/container.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+# move into base directory, in case this script is not executed via `make container`
+cd "$(dirname "$0")/.." || exit 1
+
+BRANCH=$(git branch --show-current)
+TAG=gluon:${BRANCH}
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply scripts/container.sh:10]: Added `-` to allow characters.

**Comment by mweinelt** [comment]
Moved the `Dockerfile` into `contrib/docker` and moved to that directory before building the image. Hope that resolves that.

File: scripts/container.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# normalize branch name to reflect a valid image name
+BRANCH=$(git branch --show-current | sed 's/[^a-z0-9-]/_/ig')
+TAG=gluon:${BRANCH}
+
+if [ "$(command -v podman)" ]
+then
+	(
+		cd contrib/docker
+		podman build -t "${TAG}" .
+	)
```
Comment (NeoRaider) [inline scripts/container.sh:17]:
```suggestion
	podman build -t "${TAG}" contrib/docker
```

File: scripts/container.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	podman run -it --rm --userns=keep-id --volume="$(pwd):/gluon" "${TAG}"
+elif [ "$(command -v docker)" ]
+then
+	(
+		cd contrib/docker
+		docker build -t "${TAG}" .
+	)
```
Comment (NeoRaider) [inline scripts/container.sh:24]:
```suggestion
	docker build -t "${TAG}" contrib/docker
```

**Review by NeoRaider — CHANGES_REQUESTED** [review]
Working great, just a few things to simplify in the code

File: scripts/container.sh (hunk 3 above)
Comment (mweinelt) [reply scripts/container.sh:17]: Ah, I somewhat expected it to still look at the cwd for build context. Great if that is sufficient.

**Comment by mweinelt** [comment]
As remarked earlier by Tom, this might still need documentation.

File: docs/user/getting_started.rst [hunk 5]
```diff
@@ -40,6 +40,12 @@ freshly installed Debian Stretch system the following packages are required:
 * `time` (built-in `time` doesn't work)
 * `qemu-utils`
 
+We also provide a container environment, that already tracks all these dependencies. It quickly gets you up and running, if you already have either Docker or Podman installed locally.
```
Comment (NeoRaider) [inline docs/user/getting_started.rst:43]:
```suggestion
We also provide a container environment that already tracks all these dependencies. It quickly gets you up and running, if you already have either Docker or Podman installed locally.
```

Comment (mweinelt) [reply docs/user/getting_started.rst]: Resolved.

**Pull request merged**